### PR TITLE
[View] Wizard Widget

### PR DESCRIPTION
**Pull request by maltewae**
# TL;DR: Wizard Widget
This pull request adds a widget to easily create wizard windows in LabExT. Detailed documentation on how to use it can be found below. Tests have been added to test functionality (e.g. button clicks) rather than appearance. 

**The code does not break any existing code. It simply adds code that is not already in use.**

# How to use the widget:
## 1. Basic settings
Create a class that inherits from `Wizard`. Call the `Wizard` constructor to set the basic settings. 
```python
from LabExT.View.Controls.Wizard import Wizard

class MyWizard(Wizard):
    def __init__(self, parent):
         super().__init__(
             parent, # required
             width=800, # Default: 640
             height=600, # Default: 480
             with_sidebar=True, # Default: True
             with_error=True, # Default: True
             on_cancel=self._cancel, # not required
             on_finish=self._save, # not required
             next_button_label="Next Step" # Default: "Next Step"
             previous_button_label="Previous Step", # Default: Previous Step
             cancel_button_label="Cancel and Close", # Default: Cancel
             finish_button_label="Finish and Save" # Default: Finish
         )
    ...
```
### Explanation of the settings
- `width` and `height` sets the dimension of the wizard window.
-  `with_sidebar` activates the step overview of the wizard. A frame with a width of 200 is created to the right of the content, displaying the titles of all steps and highlighting the current step. 
- `with_error` activates the error display function of the wizard. When the wizard function `setError("Error: ...")` is called, the error is displayed in red above the buttons.
- `on_cancel` is the callback function that is called when the user closes the window or clicks Cancel. This method is **blocking**. The method is expected to return a bool. If the return value is True, the window is closed, otherwise it remains open.
- `on_finish` is the callback method that is called when the user clicks on Finish. This method is **blocking**. The method is expected to return a bool. If the return value is True, the window is closed, otherwise it remains open.
- `next_button_label`, `previous_button_label`, `cancel_button_label` and `finish_button_label ` are used to change the button labels.

**Note:** The `Wizard` class itself inherits from the Tkinter class `Toplevel.` Therefore, well-known functions such as `title` are available on `self`.

## 2. Define new step
To add new steps to the wizard, the method `add_step` is used. It is recommended to define the steps in the constructor of your wizard class.
```python
self.connection_step = self.add_step(
    builder=self._connection_step_builder, # required
    title="Stage Connection", # Default: None
    on_next=self._on_next, # not required
    on_previous=self._on_previous, # not required
    on_reload=self._on_reload, # not required
    previous_step_enabled=True, # Default: True
    next_step_enabled=True, # Default: True
    finish_step_enabled=False # Default: False
)
```
### Explanation of the settings
- `builder` is the routine that builds the wizard step, i.e. defines all Tkinter objects. A Tkinter `Frame`-object is passed to the method as the first argument. All elements should use this frame as parent. The builder method is called every time the step is displayed or the wizard is manually reloaded, i.e. it is possible to render the step conditionally based on current state.
- `title` defines an optional title for the sidebar, if this has been activated.
- `on_next` is the callback method when the user clicks on "Next step". This method is **blocking**. The method is expected to return a bool. If the return value is True, the next step is loaded, otherwise not.
- `on_previous ` is the callback method when the user clicks on "Previous step". This method is **blocking**. The method is expected to return a bool. If the return value is True, the previous step is loaded, otherwise not.
- `on_reload` is called every time the step is "built", i.e. the builder method is called. This happens during step changes or when the `__reload__` method is called manually. Exemplary use: To check the current wizard state for errors:
```python
def _check_assignment(self):
    if is_stage_assignment_valid(self._current_stage_assignment):
        self.current_step.next_step_enabled = True
        self.set_error("")
    else:
        self.current_step.next_step_enabled = False
        self.set_error("Please assign at least one stage and do not select a stage twice.")
```
- `next_step_enabled ` activates the next step button. Note: This property can also be changed after the step creation (see code above).
- `previous_step_enabled ` activates the previous step button. Note: This property can also be changed after the step creation (see code above).
- `finish_step_enabled`  activates the finish button. Note: This property can also be changed after the step creation (see code above).

## 3. Define step sequence
Use the `next_step` and `previous_step` properties of the steps to define the order. Note: The order can also be changed on the fly.
```python
# Connect Steps
self.first_step.next_step = self.second_step
self.second_step.previous_step = self.first_step
self.second_step.next_step = self.third_step
self.third_step.previous_step = self.third_step
```

## 4. Define first step
To start the wizard, the first step must be defined.
```python
self.current_step = self.first_step
```

## Miscellaneous
- Use `__reload__` to reload the wizard. The method calls the builder again and updates all button and sidebar states.
- Use the `set_error(str)` method to indicate an error. Note: To reset the error, use `set_error("")`

## Screenshots
Screenshot 2022-01-25 at 13 15 54
Screenshot 2022-01-25 at 13 15 59
Screenshot 2022-01-25 at 13 16 08

### CODE REVIEW

**Review by marcoep — CHANGES_REQUESTED** [review]
Awesome! Mainly PEP8 stuff and one question :)

I will make another PR adding the very well done description of this PR to the official documentation once this is merged!

File: LabExT/View/Controls/Wizard.py [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+            self._sidebar_label = Label(
+                self.wizard._sidebar_frame,
+                anchor="w",
+                text=title,
+                foreground=self.INACTIVE_LABEL_COLOR)
+            self._sidebar_label.pack(side=TOP, fill=X, padx=10, pady=5)
+        else:
+            self._sidebar_label = None
+
+    def activate_sidebar_label(self):
```
[the remark this answers is not in the document]
Comment (maltewae) [reply LabExT/View/Controls/Wizard.py:62]: These methods are called automatically when a new step is loaded (https://github.com/LabExT/LabExT/pull/33/files#diff-e037d36f7c89514f42d2dafdd24e0a5ae9c8435e6cbb82152b97eb6dcf234f50R257 and https://github.com/LabExT/LabExT/pull/33/files#diff-e037d36f7c89514f42d2dafdd24e0a5ae9c8435e6cbb82152b97eb6dcf234f50R260). They do not have to be called manually by the user and are called when clicking on "Next Step" and "Previous Step".

File: LabExT/Tests/Utils.py [hunk 2]
```diff
@@ -5,6 +5,24 @@
 This program is free software and comes with ABSOLUTELY NO WARRANTY; for details see LICENSE file.
 """
 
+import _tkinter
+import tkinter
+from unittest import TestCase
+
+class TKinterTestCase(TestCase):
+    def setUp(self):
+        self.root=tkinter.Tk()
```
[the remark this answers is not in the document]
Comment (maltewae) [reply LabExT/Tests/Utils.py]: Changed it 👍🏼

File: LabExT/Tests/Utils.py [hunk 3]
```diff
@@ -5,6 +5,24 @@
 This program is free software and comes with ABSOLUTELY NO WARRANTY; for details see LICENSE file.
 """
```
[the remark this answers is not in the document]
Comment (maltewae) [reply LabExT/Tests/Utils.py:7]: Added one line.

File: LabExT/View/Controls/Wizard.py [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+
+    def __init__(
+            self,
+            parent,
+            width=640,
+            height=480,
+            on_cancel=None,
+            on_finish=None,
+            withSidebar=True,
+            withError=True,
```
[the remark this answers is not in the document]
Comment (maltewae) [reply LabExT/View/Controls/Wizard.py]: Changed it, thanks for the notice 👍🏼

**Review by marcoep — APPROVED** [review]
